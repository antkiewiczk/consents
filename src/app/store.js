import { configureStore } from '@reduxjs/toolkit';
import consentReducer from '../features/consent/consentsSlice'

export const store = configureStore({
  reducer: {
    consents: consentReducer,
  },
});
