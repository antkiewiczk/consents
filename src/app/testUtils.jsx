import React from 'react'
import { render as rtlRender } from '@testing-library/react'
import { configureStore } from '@reduxjs/toolkit'
import { Provider } from 'react-redux'
// Import your own reducer
import consentsReducer from '../features/consent/consentsSlice';

function render(
    ui,
    {
        preloadedState,
        store = configureStore({ reducer: { consents: consentsReducer }, preloadedState }),
        ...renderOptions
    } = {}
) {
    function Wrapper({ children }) {
        return <Provider store={store}>{children}</Provider>
    }
    return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}

// re-export everything
export * from '@testing-library/react'
// override render method
export { render }