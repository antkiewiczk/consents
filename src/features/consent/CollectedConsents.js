import { Table, TableBody, TableCell, TableFooter, TableHead, TablePagination, TableRow } from '@mui/material';
import React, { useState, useEffect } from 'react';
import { useSelector, useDispatch } from 'react-redux';
import { fetchConsentsAsync, selectAllConsents } from './consentsSlice';
import styles from './consent.module.scss';

export function CollectedConsents() {
    const [page, setPage] = useState(0);
    const dispatch = useDispatch();
    const consents = useSelector(selectAllConsents);
    const rowsPerPage = 2;

    useEffect(() => {
        dispatch(fetchConsentsAsync());
    }, [dispatch])

    const handleChangePage = (_, newPage) => {
        setPage(newPage);
    };

    if (!consents) return null;
    return (
        <div className={styles.tableWrapper}>
            <Table>
                <TableHead>
                    <TableRow>
                        <TableCell component="th" scope="row">
                            Name
                        </TableCell>
                        <TableCell style={{ width: 160 }} align="right">
                            Email
                        </TableCell>
                        <TableCell style={{ width: 160 }} align="right">
                            Consents
                        </TableCell>
                    </TableRow>
                </TableHead>
                <TableBody>
                    {consents.slice(page * rowsPerPage, page * rowsPerPage + rowsPerPage).map((row) => (
                        <TableRow key={row.name}>
                            <TableCell scope="row">
                                {row.name}
                            </TableCell>
                            <TableCell style={{ width: 220 }} align="right">
                                {row.email}
                            </TableCell>
                            <TableCell style={{ width: 220 }} align="right">
                                {row.consents.join(', ')}
                            </TableCell>
                        </TableRow>
                    ))}
                </TableBody>
                <TableFooter>
                    <TableRow>
                        <TablePagination
                            count={consents.length}
                            rowsPerPage={2}
                            page={page}
                            onPageChange={handleChangePage}
                            rowsPerPageOptions={[]}
                        />
                    </TableRow>
                </TableFooter>
            </Table>
        </div>
    );
}
