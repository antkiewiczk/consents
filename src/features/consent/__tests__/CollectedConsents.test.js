import { render } from "../../../app/testUtils";
import * as reactRedux from 'react-redux';
import { CollectedConsents } from "../CollectedConsents";
import reducer from '../consentsSlice';
import { createStore } from "@reduxjs/toolkit";

const mockedUsedNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
    ...jest.requireActual('react-router-dom'),
    useNavigate: () => mockedUsedNavigate,
}));

const initialState = reducer({ consents: { ids: [1], entities: { 1: { id: 1, name: 'John Doe', email: 'some@email.com', consents: ['newsletter'] } } } }, { type: 'test/init' });

jest.mock("react-redux", () => ({
    ...jest.requireActual("react-redux"),
    useSelector: jest.fn(),
    useDispatch: jest.fn()
}));

describe("CollectedConsents", () => {
    beforeEach(() => {
        reactRedux.useSelector.mockImplementation(callback => {
            return callback(initialState);
        });
        reactRedux.useDispatch.mockImplementation(callback => {
            return callback({ ee: 'aa' });
        });
    });

    afterEach(() => {
        reactRedux.useSelector.mockClear();
        reactRedux.useDispatch.mockClear();
    })

    test("renders nothing when store is empty", () => {
        reactRedux.useSelector.mockImplementation(callback => {
            return callback({ consents: { ids: [], entities: {} } }, { type: 'test/init' });
        });

        const dispatch = jest.fn();
        reactRedux.useDispatch.mockImplementation(() => dispatch);
        const { container } = render(
            <CollectedConsents />
        );

        expect(container).toMatchSnapshot();
    });

    test("renders correctly", () => {
        const dispatch = jest.fn();
        reactRedux.useDispatch.mockImplementation(() => dispatch);
        const { container } = render(
            <reactRedux.Provider store={createStore(() => initialState)}>
                <CollectedConsents />
            </reactRedux.Provider>
        );

        expect(container).toMatchSnapshot();
    });

    // test("validates correct input", async () => {
    //     const { getByLabelText, getByPlaceholderText, getByText } = render(
    //         <GiveConsent />
    //     );

    //     await userEvent.type(getByPlaceholderText('Name'), 'John Doe', { delay: 1 });
    //     await userEvent.type(getByPlaceholderText('Email address'), 'john@doe.com', { delay: 1 });
    //     await userEvent.click(getByLabelText('Receive newsletter'), { delay: 1 });
    //     expect(getByText('Give consent')).not.toHaveAttribute('disabled');
    // });

    // test('dispatches action when Give Consent button is clicked', async () => {
    //     const dispatch = jest.fn();
    //     reactRedux.useDispatch.mockImplementation(() => dispatch);
    //     const { getByLabelText, getByPlaceholderText, getByText } = render(
    //         <GiveConsent />
    //     );

    //     await userEvent.type(getByPlaceholderText('Name'), 'John Doe', { delay: 1 });
    //     await userEvent.type(getByPlaceholderText('Email address'), 'john@doe.com', { delay: 1 });
    //     await userEvent.click(getByLabelText('Receive newsletter'), { delay: 1 });

    //     userEvent.click(getByText('Give consent'));
    //     expect(dispatch).toBeCalledTimes(1);
    // });
})
