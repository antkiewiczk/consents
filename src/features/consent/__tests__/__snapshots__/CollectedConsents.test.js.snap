// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CollectedConsents renders correctly 1`] = `
<div>
  <div
    class="tableWrapper"
  >
    <table
      class="MuiTable-root css-rqglhn-MuiTable-root"
    >
      <thead
        class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="row"
          >
            Name
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1azl6jz-MuiTableCell-root"
            scope="col"
            style="width: 160px;"
          >
            Email
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1azl6jz-MuiTableCell-root"
            scope="col"
            style="width: 160px;"
          >
            Consents
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            scope="row"
          >
            John Doe
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
            style="width: 220px;"
          >
            some@email.com
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
            style="width: 220px;"
          >
            newsletter
          </td>
        </tr>
      </tbody>
      <tfoot
        class="MuiTableFooter-root css-9cgc89-MuiTableFooter-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-footer css-1q1u3t4-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-footer MuiTableCell-sizeMedium MuiTablePagination-root css-fikjyc-MuiTableCell-root-MuiTablePagination-root"
            colspan="1000"
          >
            <div
              class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-11bfvty-MuiToolbar-root-MuiTablePagination-toolbar"
            >
              <div
                class="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
              />
              <p
                class="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
              >
                1–1 of 1
              </p>
              <div
                class="MuiTablePagination-actions"
              >
                <button
                  aria-label="Go to previous page"
                  class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                  disabled=""
                  tabindex="-1"
                  title="Go to previous page"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="KeyboardArrowLeftIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
                    />
                  </svg>
                </button>
                <button
                  aria-label="Go to next page"
                  class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                  disabled=""
                  tabindex="-1"
                  title="Go to next page"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="KeyboardArrowRightIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </td>
        </tr>
      </tfoot>
    </table>
  </div>
</div>
`;

exports[`CollectedConsents renders nothing when store is empty 1`] = `
<div>
  <div
    class="tableWrapper"
  >
    <table
      class="MuiTable-root css-rqglhn-MuiTable-root"
    >
      <thead
        class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="row"
          >
            Name
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1azl6jz-MuiTableCell-root"
            scope="col"
            style="width: 160px;"
          >
            Email
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1azl6jz-MuiTableCell-root"
            scope="col"
            style="width: 160px;"
          >
            Consents
          </th>
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
      />
      <tfoot
        class="MuiTableFooter-root css-9cgc89-MuiTableFooter-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-footer css-1q1u3t4-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-footer MuiTableCell-sizeMedium MuiTablePagination-root css-fikjyc-MuiTableCell-root-MuiTablePagination-root"
            colspan="1000"
          >
            <div
              class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular MuiTablePagination-toolbar css-11bfvty-MuiToolbar-root-MuiTablePagination-toolbar"
            >
              <div
                class="MuiTablePagination-spacer css-1psng7p-MuiTablePagination-spacer"
              />
              <p
                class="MuiTablePagination-displayedRows css-levciy-MuiTablePagination-displayedRows"
              >
                0–0 of 0
              </p>
              <div
                class="MuiTablePagination-actions"
              >
                <button
                  aria-label="Go to previous page"
                  class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                  disabled=""
                  tabindex="-1"
                  title="Go to previous page"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="KeyboardArrowLeftIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M15.41 16.09l-4.58-4.59 4.58-4.59L14 5.5l-6 6 6 6z"
                    />
                  </svg>
                </button>
                <button
                  aria-label="Go to next page"
                  class="MuiButtonBase-root Mui-disabled MuiIconButton-root Mui-disabled MuiIconButton-colorInherit MuiIconButton-sizeMedium css-zylse7-MuiButtonBase-root-MuiIconButton-root"
                  disabled=""
                  tabindex="-1"
                  title="Go to next page"
                  type="button"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="KeyboardArrowRightIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M8.59 16.34l4.58-4.59-4.58-4.59L10 5.75l6 6-6 6z"
                    />
                  </svg>
                </button>
              </div>
            </div>
          </td>
        </tr>
      </tfoot>
    </table>
  </div>
</div>
`;
